Build-process.md

```diff
@@ -0,0 +1,114 @@
+## Table of contents
+
+* [Introduction](#introduction)
+* [Build script](#build-script)
+* [Build modes](#build-modes)
+  * [Build mode constants](#build-mode-constants)
+  * [Dev mode](#dev-mode)
+  * [Prod mode](#prod-mode)
+    * [Plain prod mode](#plain-prod-mode)
+    * [Minify only third-party libraries](#minify-only-third-party-libraries)
+    * [Maintenance mode](#maintenance-mode)
+    * [Deploy mode](#deploy-mode)
+    * [Unused: Deparallelized terser mode](#unused-deparallelized-terser-mode)
+* [Source maps](#source-maps)
+
+## Introduction
+
+Before we can run the Oppia server, we have to build it. This build process transforms the files defining our application (e.g. code, images, configuration files, etc.) from the developer-friendly format we store in GitHub into a form that can be executed to run the server. If you are running the local development server, this build process makes minimal changes, but when deploying to production, we do much more optimization.
+
+This page documents how the build process works in each of its modes. For most developers the build steps are handled automatically and can be ignored; however, some bugs appear only in certain build modes. For example, a bug might only happen in prod mode, in which case a developer will be unable to reproduce it in dev mode.
+
+## Build script
+
+The build process is performed by [`scripts/build.py`](https://github.com/oppia/oppia/blob/develop/scripts/build.py). If you ever want to know exactly how a particular part of the build process works, you should check out the code there.
+
+## Build modes
+
+There are two primary build modes: dev mode and prod mode. When running in prod mode, you can also enable maintenance mode or deploy mode, or you can choose to only minify third-party libraries. This means the build mode hierarchy looks like this:
```

Comment:
Are the three prod mode categories mutually exclusive? It's not clear because the third thing isn't a "mode" and sounds like an add-on, but the structure below suggests that they are mutually exclusive things.

EDIT/UPDATE: I saw the sentence after this list. Maybe we should think of these in terms of flags/options rather than modes? The question of which flags are mutually exclusive (if any) still needs to be explained/described, though. It would also be nice to rename things both here and in the codebase so that we don't use "mode" for things that aren't modes.

Reply:
I've added a list of all allowed mode combinations to clarify which are mutually exclusive. I agree with renaming the modes, and I've opened #13909 to get @vojtechjelinek's input